doc: Add federatedhpa user guide doc (#368)
karmada-io/website: docs/userguide/autosscaling/autoscaling-with-federatedhpa.md

---
title: Autoscaling across clusters with FederatedHPA
---

> For workload deployed across multiple clusters that require autoscaling to handle varying requests, Karmada FederatedHPA offers the ability to scale the workload across clusters and the following benefits:

Reviewer: ```suggestion
For workloads deployed across multiple clusters that require autoscaling to handle varying requests, Karmada's FederatedHPA offers the ability to scale the workload across clusters and brings the following benefits:
```

### The ServiceExport and ServiceImport CRDs have been installed

We need to install ServiceExport and ServiceImport in the member clusters.

After ServiceExport and ServiceImport have been installed on the **Karmada Control Plane**, we can create `ClusterPropagationPolicy` to propagate those two CRDs to the member clusters.

```yaml
# propagate ServiceExport CRD
apiVersion: policy.karmada.io/v1alpha1
kind: ClusterPropagationPolicy
metadata:
  name: serviceexport-policy
spec:
  resourceSelectors:
    - apiVersion: apiextensions.k8s.io/v1
      kind: CustomResourceDefinition
      name: serviceexports.multicluster.x-k8s.io
  placement:
    clusterAffinity:
      clusterNames:
        - member1
        - member2
---        
# propagate ServiceImport CRD
apiVersion: policy.karmada.io/v1alpha1
kind: ClusterPropagationPolicy
metadata:
  name: serviceimport-policy
spec:
  resourceSelectors:
    - apiVersion: apiextensions.k8s.io/v1
      kind: CustomResourceDefinition
      name: serviceimports.multicluster.x-k8s.io
  placement:
    clusterAffinity:
      clusterNames:
        - member1
        - member2
```

### metrics-server has been installed in member clusters

> We need to install `metrics-server` for member clusters, install it by running `hack/deploy-k8s-metrics-server.sh ${member_cluster_kubeconfig} ${member_cluster_context_name}` directly.  

Reviewer: How about putting the following into a single line? It seems a little crowded.
```
`hack/deploy-k8s-metrics-server.sh ${member_cluster_kubeconfig} ${member_cluster_context_name}` directly.
```

```yaml
apiVersion: autoscaling.karmada.io/v1alpha1
kind: FederatedHPA
metadata:
  name: nginx
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: nginx
  minReplicas: 1
  maxReplicas: 10
  behavior:
    scaleDown:
      stabilizationWindowSeconds: 10
    scaleUp:
      stabilizationWindowSeconds: 10
  metrics:
    - type: Resource
      resource:
        name: cpu
        target:
          type: Utilization
          averageUtilization: 10
```

### Step 3: Export service to `member1` cluster

* Create a `ServiceExport` object on Karmada Control Plane, and then create a PropagationPolicy to propagate the ServiceExport object to `member1` and `member2` cluster.
  ```yaml
  apiVersion: multicluster.x-k8s.io/v1alpha1
  kind: ServiceExport
  metadata:
    name: nginx-service
  ---
  apiVersion: policy.karmada.io/v1alpha1
  kind: PropagationPolicy
  metadata:
    name: serve-export-policy
  spec:
    resourceSelectors:
      - apiVersion: multicluster.x-k8s.io/v1alpha1
        kind: ServiceExport
        name: nginx-service
    placement:
      clusterAffinity:
        clusterNames:
          - member1
          - member2
  ```

> * Create a `ServiceImport` object on Karmada Control Plane, and then create a PropagationPolicy to propagate the ServiceImport object to `member1` cluster.

Reviewer: ```suggestion
* Create a `ServiceImport` object on Karmada Control Plane, and then create a `PropagationPolicy` to propagate the `ServiceImport` object to `member1` cluster.
```

kind: Deployment
metadata:
  name: hey
  labels:
    app: hey
spec:
  replicas: 1
  selector:
    matchLabels:
      app: hey
  template:
    metadata:
      labels:
        app: hey
    spec:
      containers:
      - image: cesign/hey:latest
        name: hey
        args:
        - sleep
        - "1000000"
---
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: nginx-propagation
spec:
  resourceSelectors:
    - apiVersion: apps/v1
      kind: Deployment
      name: hey
  placement:
    clusterAffinity:
      clusterNames:
        - member1
```

### Step 5: Test scaling up

* Check the pod distribution firstly.
  ```sh
  $ karmadactl get pods
  NAME                     CLUSTER   READY   STATUS      RESTARTS   AGE
  hey-79f8658b77-vdk2t     member1   1/1     Running     0          67m
  nginx-777bc7b6d7-mbdn8   member1   1/1     Running     0          61m
  ```

* Check service ip of the nginx service.
  ```sh
  $ karmadactl get svc
  NAME                    CLUSTER   TYPE        CLUSTER-IP        EXTERNAL-IP   PORT(S)   AGE   ADOPTION
  derived-nginx-service   member1   ClusterIP   10.11.59.213      <none>        80/TCP    20m   Y
  ```

> * Requesting nginx service in hey pod to increase the nginx pods' CPU usage.

Reviewer: ```suggestion
* Request nginx service in hey pod to increase the nginx pods' CPU usage.
```

---
title: Autoscaling across clusters with FederatedHPA
---

For workloads deployed across multiple clusters that require autoscaling to handle varying requests, Karmada's FederatedHPA offers the ability to scale the workload across clusters and brings the following benefits:
* Coordinate resources from multiple clusters, scale workload across clusters, and improve overall resource utilization of multiple clusters.
* Config cross-cluster autoscaling uniformly and efficiently.
* Avoid single point of failure causing server downtime, cross-cluster autoscaling improves workload stability.
* Limit resource usage across clusters uniformly.

The overall architecture is as follows:  
<img src="../../resources/userguide/autoscaling/federatedhpa-native-metrics.png" width="60%">

> 1. HPA controller query metrics via metrics API `metrics.k8s.io` with label selector periodically.

Reviewer: ```suggestion
1. HPA controller queries metrics via metrics API `metrics.k8s.io` with label selector periodically.
```

> 1. `karmada-apiserver` get the metrics API query, and it will route to karmada-metrics-adapter via API service registration.

Reviewer: ```suggestion
1. `karmada-apiserver` gets the metrics API query, and it will route to karmada-metrics-adapter via API service registration.
```

1. `karmada-metrics-adapter` will query the metrics from the target clusters(where the pod exists). After the metrics are collected, it will aggregate them and return it.
1. HPA controller will calculate the desired replicas based on metrics and scale the workload directly. Then `karmada-scheduler` will schedule the replicas to the member clusters.

> Note: To use this feature, The Karmada version must be v1.6.0 or later.

## Prerequisites

### Karmada has been installed

We can install Karmada by referring to [Quick Start](https://github.com/karmada-io/karmada#quick-start), or directly run `hack/local-up-karmada.sh` script which is also used to run our E2E cases.

### Member Cluster Network

Ensure that at least two clusters have been added to Karmada, and the container networks between member clusters are connected.

- If you use the `hack/local-up-karmada.sh` script to deploy Karmada, Karmada will have three member clusters, and the container networks of the `member1` and `member2` will be connected.

> - You can use `Submariner` or other related open source projects to connected networks between member clusters.

Reviewer: ```suggestion
- You can use `Submariner` or other related open source projects to connect networks between member clusters.
```
